### PR TITLE
Fixed several issues related to creating tasks with cloud data

### PR DESCRIPTION
**Pull request by Marishka17**
### Motivation and context

@coderabbitai summary
### Checklist

- [x] I submit my changes into the `develop` branch
- [x] I have created a changelog fragment 
- [ ] I have updated the documentation accordingly
- [x] I have added tests to cover my changes (*partially*)
- [ ] I have linked related issues (see [GitHub docs](
  https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword))
- [ ] I have increased versions of npm packages if it is necessary
  ([cvat-canvas](https://github.com/cvat-ai/cvat/tree/develop/cvat-canvas#versioning),
  [cvat-core](https://github.com/cvat-ai/cvat/tree/develop/cvat-core#versioning),
  [cvat-data](https://github.com/cvat-ai/cvat/tree/develop/cvat-data#versioning) and
  [cvat-ui](https://github.com/cvat-ai/cvat/tree/develop/cvat-ui#versioning))

### License

- [x] I submit _my code changes_ under the same [MIT License](
  https://github.com/cvat-ai/cvat/blob/develop/LICENSE) that covers the project.
  Feel free to contact the maintainers if that's a concern.

## Summary by CodeRabbit

- **New Features**
  - Improved media download performance with parallel downloading.
  - Enhanced file handling with the new `NamedBytesIO` class.
  - Added support for specifying stop frames in task manifest generation.
  - Enhanced `DatasetImagesReader` to handle generator sources.

- **Performance Improvements**
  - Optimized image download methods to use threading for faster processing.

- **Configuration**
  - Introduced new settings for maximum threads and files per thread in cloud data downloading.

These updates enhance the flexibility, performance, and configurability of media handling and downloading in the application.

### CODE REVIEW

File: cvat/apps/engine/task.py [hunk 1]
```diff
@@ -482,10 +482,19 @@ def _create_task_manifest_from_cloud_data(
...
     cloud_storage_instance = db_storage_to_storage_instance(db_storage)
-    content = cloud_storage_instance.bulk_download_to_memory(sorted_media)
-    manifest.link(sources=content, DIM_3D=dimension == models.DimensionType.DIM_3D)
+    content_generator = cloud_storage_instance.bulk_download_to_memory(sorted_media)
+
+    manifest.link(
+        sources=content_generator,
+        DIM_3D=dimension == models.DimensionType.DIM_3D,
+        stop=stop_frame,
+    )
```
Comment (zhiltsov-max) [inline cvat/apps/engine/task.py:497]: Did you check if the reverse order of results, returned by the generator, is not a problems here?

Comment (Marishka17) [reply cvat/apps/engine/task.py:497]: Why do you think that results will be reversed? Yes, I've checked that after using different sorting methods tasks will be created with the correct, expected file order. I guess we have such tests but need to check.

Comment (zhiltsov-max) [reply cvat/apps/engine/task.py:497]: You can see it in [my previous comment](https://github.com/cvat-ai/cvat/pull/7903#discussion_r1615919002) with the small script output - the item numbers are reversed. You can also find it in the CPython implementation [here](https://github.com/python/cpython/blob/48f9d3e3faec5faaa4f7c9849fecd27eae4da213/Lib/concurrent/futures/_base.py#L616-L617).

Comment (Marishka17) [reply cvat/apps/engine/task.py:497]: Yes, I also checked the source code before asking you but you can see that after reversing the list with futures, the next future will be the last one from the reversed list that corresponds to the first one from the original list. So I still don't understand..

Comment (zhiltsov-max) [reply cvat/apps/engine/task.py:497]: Ok, I investigated it one more time, it seems to be the detail of the list call.

**Review by zhiltsov-max — APPROVED** [review]

File: cvat/apps/engine/task.py (hunk 1 above)
Comment (Marishka17) [reply cvat/apps/engine/task.py:497]: @zhiltsov-max, I've added REST API tests to check task creation using cloud data and different sorting methods.

File: tests/python/rest_api/test_tasks.py [hunk 2]
```diff
@@ -1646,19 +1654,24 @@ def test_create_task_with_file_pattern(
...
     ):
         cloud_storage = cloud_storages[cloud_storage_id]
         task_id, _ = self._create_task_with_cloud_data(
             request=request,
             cloud_storage=cloud_storage,
-            use_manifest=True,
-            use_cache=True,
-            server_files=[],
+            # manifest file should not be uploaded if random sorting is used or if cache is not used
+            use_manifest=use_manifest & use_cache & (sorting_method != "random"),
```
Comment (zhiltsov-max) [inline tests/python/rest_api/test_tasks.py:1670]: Why not use `and` for this?

**Review by zhiltsov-max — APPROVED** [review]

**Pull request merged**